src/lib/libc/gen/times.3
.\"	times.3,v 1.15 2003/08/07 16:42:58 agc Exp
.\"
.\" Copyright (c) 1990, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)times.3	8.1 (Berkeley) 6/4/93
.\"
.Dd June 4, 1993
.Dt TIMES 3
.Os
.Sh NAME
.Nm times
.Nd process times
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In sys/times.h
.Ft clock_t
.Fn times "struct tms *tp"
.Sh DESCRIPTION
.Bf -symbolic
This interface is obsoleted by
.Xr getrusage 2
and
.Xr gettimeofday 2 .
.Ef
.Pp
The
.Fn times
function returns the value of time in clock ticks since 0 hours, 0
minutes, 0 seconds, January 1, 1970, Coordinated Universal Time (UTC).
.Pp
The number of clock ticks per second may be determined by calling
.Xr sysconf 3
with the
.Dv _SC_CLK_TCK
request.
It is generally (but not always) between 60 and 1024.
.Pp
Note that at the common rate of 100 ticks per second on many
.Nx
ports, and with a 32-bit unsigned clock_t, this value first wrapped in 1971.
.Pp
The
.Fn times
call also fills in the structure pointed to by
.Fa tp
with time-accounting information.
.Pp
The
.Fa tms
structure is defined as follows:
.Bd -literal -offset indent
typedef struct {
	clock_t tms_utime;
	clock_t tms_stime;
	clock_t tms_cutime;
	clock_t tms_cstime;
}
.Ed
.Pp
The elements of this structure are defined as follows:
.Bl -tag -width tms_cutime
.It Fa tms_utime
The
.Tn CPU
time charged for the execution of user instructions.
.It Fa tms_stime
The
.Tn CPU
time charged for execution by the system on behalf of
the process.
.It Fa tms_cutime
The sum of the
.Fa tms_utime  s
and
.Fa tms_cutime  s
of the child processes.
.It Fa tms_cstime
The sum of the
.Fa tms_stime Ns s
and
.Fa tms_cstime Ns s
of the child processes.
.El
.Pp
All times are measured in clock ticks, as defined above.
Note that at 100 ticks per second,
and with a 32-bit unsigned clock_t,
the values wrap after 497 days.
.Pp
The times of a terminated child process are included in the
.Fa tms_cutime
and
.Fa tms_cstime
elements of the parent when one of the
.Xr wait 2
functions returns the process ID of the terminated child to the parent.
If an error occurs,
.Fn times
returns the value
.Pq (clock_t)\-1 ,
and sets
.Va errno
to indicate the error.
.Sh ERRORS
The
.Fn times
function
may fail and set the global variable
.Va errno
for any of the errors specified for the library
routines
.Xr getrusage 2
and
.Xr gettimeofday 2 .
.Sh SEE ALSO
.Xr time 1 ,
.Xr getrusage 2 ,
.Xr gettimeofday 2 ,
.Xr wait 2 ,
.Xr sysconf 3
.Sh STANDARDS
The
.Fn times
function conforms to
.St -p1003.1-90 .


src/lib/libc/gen/times.c
/*	times.c,v 1.15 2009/01/11 02:46:27 christos Exp	*/

/*-
 * Copyright (c) 1990, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#if defined(LIBC_SCCS) && !defined(lint)
#if 0
static char sccsid[] = "@(#)times.c	8.1 (Berkeley) 6/4/93";
#else
__RCSID("times.c,v 1.15 2009/01/11 02:46:27 christos Exp");
#endif
#endif /* LIBC_SCCS and not lint */

#include "namespace.h"
#include <sys/param.h>
#include <sys/time.h>
#include <sys/times.h>
#include <sys/resource.h>

#include <assert.h>
#include <errno.h>
#include <time.h>

/*
 * Convert usec to clock ticks; could do (usec * CLK_TCK) / 1000000,
 * but this would overflow if we switch to nanosec.
 */
#define	CONVTCK(r)	\
    (clock_t)(r.tv_sec * clk_tck + r.tv_usec / (1000000 / (uint)clk_tck))

#ifndef __times_rusage
#define __times_rusage struct rusage
#endif
#ifndef __times_timeval
#define __times_timeval struct timeval
#endif

clock_t
times(struct tms *tp)
{
	__times_rusage ru;
	__times_timeval t;
	static clock_t clk_tck;
	
	_DIAGASSERT(tp != NULL);

	/*
	 * we use a local copy of CLK_TCK because it expands to a
	 * moderately expensive function call.
	 */
	if (clk_tck == 0)
		clk_tck = (clock_t)CLK_TCK;

	if (getrusage(RUSAGE_SELF, &ru) < 0)
		return ((clock_t)-1);
	tp->tms_utime = CONVTCK(ru.ru_utime);
	tp->tms_stime = CONVTCK(ru.ru_stime);
	if (getrusage(RUSAGE_CHILDREN, &ru) < 0)
		return ((clock_t)-1);
	tp->tms_cutime = CONVTCK(ru.ru_utime);
	tp->tms_cstime = CONVTCK(ru.ru_stime);
	if (gettimeofday(&t, (struct timezone *)0))
		return ((clock_t)-1);
	return ((clock_t)(CONVTCK(t)));
}
